pkg/mover/chart.go

```diff
@@ -236,17 +276,127 @@ func namespacedPath(fullpath, chartName string) string {
+
+A move to tar will:
+
+1 - Drop all images locally
+
+2 - Pack also the original chart (unpacked) and the hints file
+
+3 - Package all in a single compressed tarball
 */
 func (cm *ChartMover) Move() error {
+	if cm.targetOfflineTar != "" {
+		return cm.archive()
+	}
+	return cm.moveChart()
+}
+
+func (cm *ChartMover) archive() error {
 	log := cm.logger
+	log.Printf("Archiving %s@%s...\n", cm.chart.Name(), cm.chart.Metadata.Version)
 
+	tarFolder := cm.targetOfflineTar + ".folder"
+	if err := os.MkdirAll(tarFolder, DefaultTarPermissions); err != nil {
+		return fmt.Errorf("failed to create tar folder %s: %w", tarFolder, err)
+	}
+	if err := archiveChart(tarFolder, cm.chartOrigin); err != nil {
+		return fmt.Errorf("failed archiving chart %s: %w", cm.chart.Name(), err)
+	}
+	cm.logger.Println("Archived char tarball")
+	if err := archiveImages(tarFolder, cm.imageChanges, cm.logger); err != nil {
+		return fmt.Errorf("failed archiving images: %w", err)
+	}
+	if cm.externalHintsFile == "" {
+		return nil
+	}
+	log.Printf("Inserting hints file %s...\n", cm.externalHintsFile)
+	archivedHints := filepath.Join(tarFolder, ArchivedHintsFilename)
+	return copyFile(cm.externalHintsFile, archivedHints)
+}
+
+func archiveChart(folder, chartPath string) error {
+	target := filepath.Join(folder, filepath.Base(chartPath))
+	if err := os.Mkdir(target, DefaultTarPermissions); err != nil {
+		return err
+	}
+	return copyRecursive(chartPath, target)
+}
+
+func copyRecursive(source, destination string) error {
+	return filepath.Walk(source, func(path string, info os.FileInfo, err error) error {
+		destPath := filepath.Join(destination, path[len(source):])
+		if path == source {
+			return nil
+		}
+		if info.IsDir() {
+			return os.Mkdir(destPath, DefaultTarPermissions)
+		}
+		return copyFile(path, destPath)
+	})
+}
+
+func copyFile(source, destination string) error {
+	sourceFileStat, err := os.Stat(source)
+	if err != nil {
+		return err
+	}
+	if !sourceFileStat.Mode().IsRegular() {
+		return fmt.Errorf("%s is not a regular file", source)
+	}
+	in, err := os.Open(source)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+	out, err := os.Create(destination)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	_, err = io.Copy(out, in)
+	return err
+}
+
+func archiveImages(folder string, imageChanges []*internal.ImageChange, logger Logger) error {
+	imagesFolder := filepath.Join(folder, "images")
+	if err := os.MkdirAll(imagesFolder, DefaultTarPermissions); err != nil {
+		return err
+	}
+	for _, change := range imageChanges {
+		name := path.Base(change.ImageReference.Context().Name())
+		tag := change.ImageReference.Identifier()
+		imageName := fmt.Sprintf("%s-%s", name, tag)
+		logger.Printf("Archiving %s...\n", imageName)
+		imageTarget := filepath.Join(imagesFolder, fmt.Sprintf("%s.tar", imageName))
+		if err := archiveImage(imageTarget, change.ImageReference, change.Image); err != nil {
+			return err
+		}
```

Comment:
Instead of saving each image individually we might want to check this MultiWrite instead:
https://pkg.go.dev/github.com/google/go-containerregistry@v0.5.1/pkg/v1/tarball#MultiWrite

I reckon many chart images might have common base layers we can avoid storing several times in the same offline archive. That might replace the whole loop.

Reply:
Done